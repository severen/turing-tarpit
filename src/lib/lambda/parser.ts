/** @license
 * SPDX-FileCopyrightText: 2022 Severen Redwood <me@severen.dev>
 * SPDX-FileCopyrightText: 2022 Keetley Rate <fridgeineyes@gmail.com>
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import { assert } from "$lib/assert";
import {
  mkAbs,
  mkApp,
  mkBVar,
  mkFVar,
  type Abs,
  type BVar,
  type FVar,
  type Term,
} from "$lib/lambda/syntax";

// In EBNF notation, we parse the following concrete syntax for λ-terms:
//   <term> ::= <abstraction> | <application>
//   <abstraction> ::= "\\" <identifier>+ "->" <term>
//   <application> ::= <application> <atom> | <atom>
//   <atom> ::= <variable> | "(" <term> ")"
//   <variable> ::= <identifier>
//
// Although the above is the classical named representation of λ-terms, we use a _locally
// nameless_ internal representation, which is described in $lib/lambda/syntax.ts.

/** A syntax error encountered while parsing a λ-calculus program. */
export class SyntaxError extends Error {
  /** The start location of this syntax error within the input. */
  position: number;

  constructor(position: number, message: string) {
    super(message);
    this.name = "SyntaxError";
    this.position = position;
  }
}

/** Parse a λ-calculus term. */
export function parse(input: string): Term {
  return new Parser(input).parse();
}

/** A parser for λ-calculus programs. */
class Parser {
  /** The input string to parse. */
  #input: string;
  /** The current character position of this parser within the input string. */
  #position = 0;

  /**
   * The stack of currently bound variables, where the 'front' of the stack is considered to be index 0
   *
   * Note that the length of this stack corresponds to the current _binding depth_.
   */
  #bound: string[] = [];

  readonly #reIdent = /^\p{ID_Start}\p{ID_Continue}*'?/u;
  readonly #reWhitespace = /^\p{Pattern_White_Space}+/u;

  /**
   * Construct a new λ-calculus parser.
   * @param input The input string to parse as a λ-calculus program.
   */
  constructor(input: string) {
    this.#input = input.trim();
  }

  /** Parse the input into a λ-calculus term. */
  parse(): Term {
    if (this.#input.length === 0) {
      throw new SyntaxError(this.#position, "unexpected EOF");
    }

    assert(!this.#isAtEnd(), "parser exhausted");

    const t = this.#term();
    if (!this.#isAtEnd()) {
      throw new SyntaxError(this.#position, `expected EOF, got ${this.#peek()}`);
    }
    return t;
  }

  /** Parse the term nonterminal. */
  #term(): Term {
    return this.#peek() === "\\" ? this.#abs() : this.#app();
  }

  /** Parse the abstraction nonterminal. */
  #abs(): Abs {
    const isIdentPrefix = (char: string): boolean => this.#reIdent.test(char);

    this.#consume("\\");
    const head = [];
    while (isIdentPrefix(this.#peek())) {
      const name = this.#ident();
      head.push(name);
    }
    // We reverse the head and push to the front to mantain an ordering such that the
    // most recently bound variables recieve a lower index.
    this.#bound.unshift(...head.slice().reverse());

    this.#consume("->");

    const body = this.#term();
    head.forEach(() => this.#bound.shift());

    // Since the internal abstract syntax has no notion of multivariate abstraction, we
    // desugar the multivariate abstraction into a sequence of nested single-variable
    // abstractions.
    return head
      .slice(0, head.length - 1)
      .reduceRight(
        (body, head) => mkAbs(head, body),
        mkAbs(head[head.length - 1], body),
      );
  }

  /** Parse the application nonterminal. */
  #app(): Term {
    const isAtomPrefix = (char: string): boolean =>
      char === "(" || this.#reIdent.test(char);

    let lhs = this.#atom();
    while (isAtomPrefix(this.#peek())) {
      const rhs = this.#atom();
      lhs = mkApp(lhs, rhs);
    }

    return lhs;
  }

  /** Parse the atom nonterminal. */
  #atom(): Term {
    if (this.#peek() !== "(") {
      return this.#var();
    }

    this.#consume("(");
    const t = this.#term();
    this.#consume(")");

    return t;
  }

  /** Parse the variable nonterminal. */
  #var(): FVar | BVar {
    const name = this.#ident();
    return !this.#bound.includes(name)
      ? // If we have reached this variable without it being bound by an abstraction,
        // then it _must_ be a free variable.
        mkFVar(name)
      : // Otherwise, it is bound, so we consult the binding stack to determine its de
        // Bruijn index.
        mkBVar(this.#bound.indexOf(name));
  }

  /** Parse an identifier. */
  #ident(): string {
    const matches = this.#input.substring(this.#position).match(this.#reIdent);
    if (!matches) {
      throw new SyntaxError(this.#position, `expected identifier, got ${this.#peek()}`);
    }

    const ident = matches[0];
    this.#consume(ident);
    return ident;
  }

  /** Check if the parser has consumed the entire input. */
  #isAtEnd(): boolean {
    return this.#position === this.#input.length;
  }

  /**
   * Get the character at the current position of the parser, or the null character if
   * the entire input has been consumed.
   */
  #peek(): string {
    return !this.#isAtEnd() ? this.#input[this.#position] : "\0";
  }

  /** Assert the presence of a lexeme in the input and advance the parser. */
  #consume(lexeme: string): void {
    const start = this.#position;
    const end = start + lexeme.length;
    const substring = this.#input.substring(start, end);
    if (substring !== lexeme) {
      throw new SyntaxError(this.#position, `expected ${lexeme}, got ${substring}`);
    }

    this.#position += lexeme.length;
    this.#discardWhitespace();
  }

  /** Advance the parser over whitespace in the input. */
  #discardWhitespace(): void {
    const matches = this.#input.substring(this.#position).match(this.#reWhitespace);
    if (matches) {
      this.#position += matches[0].length;
    }
  }
}
