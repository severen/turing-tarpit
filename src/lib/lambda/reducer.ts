/** @license
 * SPDX-FileCopyrightText: 2022 Severen Redwood <me@severen.dev>
 * SPDX-FileCopyrightText: 2022 Keetley Rate <fridgeineyes@gmail.com>
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import {
  freeVars,
  mkAbs,
  mkApp,
  mkVar,
  TermKind,
  type Abs,
  type Name,
  type Term,
} from "$lib/lambda/syntax";

/**
 * Perform reduction on a term until no more reductions are possible.
 * @param t The term to evaluate.
 * @returns The reduction steps taken to evaluate `t`.
 */
export function evaluate(t: Term): Term[] {
  const steps = [t, reduce(t)[0]];
  for (let i = 1; ; ++i) {
    const [s, wasContracted] = reduce(steps[i]);

    // If reduction gives the same term twice, then there are no more β-redexes and we
    // can stop evaluating. Also note the use of isEqual from lodash over ===, which we
    // use for _structural_ equality instead of _referential_ equality. One would expect
    // JavaScript to have such a function built in, but unfortunately it still sucks.
    if (wasContracted) {
      steps.push(s);
    } else {
      return steps;
    }
  }
}

// TODO: Profile this to see just how slow it is in comparison to my implementation in
//       Haskell (severen/sly).
/**
 * Perform one step of normal order β-reduction within a term, if possible.
 * @param t The term in which to perform β-reduction.
 * @returns A list R in which R[0] is the resultant term and R[1] indicates whether a
 * contraction has occured.
 */
export function reduce(t: Term): [Term, boolean] {
  // So that we can enforce leftmost-to-rightmost evaluation of redexes, we keep track
  // of whether a contraction has ocurred with the use of the second parameter.
  switch (t.kind) {
    case TermKind.Var:
      return [t, false];
    case TermKind.Abs: {
      const [body, wasContracted] = reduce(t.body);
      return [mkAbs(t.head, body), wasContracted];
    }
    case TermKind.App: {
      if (t.left.kind === TermKind.Abs) {
        return [subst(t.left.head, t.right, t.left.body), true];
      }

      // If the left-hand term of the application has not been contracted, then we
      // try to reduce the right-hand term instead.
      const [left, wasLeftContracted] = reduce(t.left);
      const [right, wasRightContracted] = wasLeftContracted
        ? [t.right, false]
        : reduce(t.right);
      return [mkApp(left, right), wasLeftContracted || wasRightContracted];
    }
  }
}

/**
 * Replace all free occurences of the variable `x` in a term `t` with the term `s`.
 *
 * To avoid variable capture, bound variables are renamed when necessary.
 *
 * @param x The variable to replace with `s`.
 * @param s The term to replace `x`.
 * @param t The term in which to perform the replacement `x -> s`.
 */
export function subst(x: Name, s: Term, t: Term): Term {
  const fvs = freeVars(s);

  const go = (t: Term): Term => {
    switch (t.kind) {
      case TermKind.Var:
        if (x === t.name) {
          return s;
        }
        return t;
      case TermKind.Abs: {
        if (x === t.head) {
          return t;
        } else if (x !== t.head && !fvs.has(t.head)) {
          return mkAbs(t.head, go(t.body));
        }
        const newHead = chooseName(t.head, new Set([t.head, ...freeVars(t.body)]));
        return go(rename(newHead, t));
      }
      case TermKind.App:
        return mkApp(go(t.left), go(t.right));
    }
  };

  return go(t);
}

/**
 * Rename the bound variable in the abstraction `f` to the name `x`, if possible.
 * @param x The new name for the bound variable of `f`.
 * @param f The abstraction on which to perform renaming.
 */
function rename(x: Name, f: Abs): Abs {
  if (!freeVars(f.body).has(x)) {
    return mkAbs(x, subst(f.head, mkVar(x), f.body));
  }

  return f;
}

/**
 * Choose a name derived from `x` that is not one of the names in `ys`.
 * @param x The name from which to derive a new name.
 * @param ys The names to avoid clashing with.
 */
function chooseName(x: Name, ys: Set<Name>): Name {
  let z = x + "'";
  for (let i = 1; ys.has(z); ++i) {
    if (i < 3) {
      z += "'";
    } else {
      z = x + i.toString();
    }
  }

  return z;
}
