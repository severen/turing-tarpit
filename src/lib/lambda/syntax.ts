/** @license
 * SPDX-FileCopyrightText: 2022 Severen Redwood <me@severen.dev>
 * SPDX-FileCopyrightText: 2022 Keetley Rate <fridgeineyes@gmail.com>
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import { assert } from "$lib/assert";

// The abstract syntax for λ-terms defined here is a _locally nameless_ representation
// similar to that defined by McBride and McKinna in 'I am not a number - I am a free
// variable' (https://doi.org/10.1145/1017472.1017477).
//
// In short, the idea is to represent free variables by names, as usual, but to represent
// _bound_ variables with _de Bruijn indices_. In order to preserve the original variable
// names from the concrete syntax, each binder continues to carry a name (taken from the
// concrete syntax) for use as a hint for the pretty printer.

/** A λ-term. */
export type Term = FVar | BVar | Abs | App;

/** The kind of a Term. */
export enum TermKind {
  /** A free variable. */
  FVar,
  /** A bound variable. */
  BVar,
  /** A λ-abstraction. */
  Abs,
  /** An application. */
  App,
}

/** A free variable term. */
export type FVar = {
  readonly kind: TermKind.FVar;
  /** The name of this variable. */
  readonly name: string;
};

/** A bound variable term. */
export type BVar = {
  readonly kind: TermKind.BVar;
  /**
   * The de Bruijn index which represents this variable.
   *
   * By _de Bruijn index_, it is meant the number of binders between this bound variable
   * and its corresponding binder. For example, `\x -> \y -> x` corresponds to `\ -> \ ->
   * 1`, in which we see that x corresponds to 1 (and not, for example, 2) due to the \y
   * binder in between \x and x.
   */
  readonly index: number;
};

/** A λ-abstraction term. */
export type Abs = {
  readonly kind: TermKind.Abs;
  /**
   * The name of the variable bound in the head of this abstraction.
   *
   * As substitution is performed with de Bruijn indices, this field is _only_ present as
   * a 'hint' for the pretty printer so that it may present symbolic names corresponding
   * to the user's original input.
   */
  readonly head: string;
  /** The body (or scope) of this abstraction. */
  readonly body: Term;
};

/** An application term. */
export type App = {
  readonly kind: TermKind.App;
  /** The left applicand. */
  readonly left: Term;
  /** The right applicand. */
  readonly right: Term;
};

/**
 * Construct a new variable term.
 * @param name The name of this variable.
 */
export const mkFVar = (name: string): FVar => ({ kind: TermKind.FVar, name });

/**
 * Construct a new variable term.
 * @param index The de Bruijn index which represents the variable.
 */
export const mkBVar = (index: number): BVar => {
  // Paranoid assertion to ensure the provided index is actually a natural
  // number.
  assert(Number.isSafeInteger(index) && index >= 0, "expected natural number");

  return { kind: TermKind.BVar, index };
};

/**
 * Construct a new λ-abstraction term.
 * @param head The name of the variable bound by the abstraction.
 * @param body The body of the abstraction.
 */
export const mkAbs = (head: string, body: Term): Abs => ({
  kind: TermKind.Abs,
  head,
  body,
});

/**
 * Construct a new application term.
 * @param left The applicand.
 * @param right The argument.
 */
export const mkApp = (left: Term, right: Term): App => ({
  kind: TermKind.App,
  left,
  right,
});

/**
 * Apply a function to each term encountered while traversing a syntax tree of
 * terms.
 * @param f The function to apply to each term in the syntax tree.
 * @param root The root term of the syntax tree.
 */
export function apply(f: (t: Term) => void, root: Term): void {
  const todo = [root];
  while (todo.length > 0) {
    // By the loop condition above, t will never be undefined/null.
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const t = todo.pop()!;
    f(t);

    if (t.kind === TermKind.App) {
      todo.push(t.right);
      todo.push(t.left);
    } else if (t.kind === TermKind.Abs) {
      todo.push(t.body);
    }
  }
}
