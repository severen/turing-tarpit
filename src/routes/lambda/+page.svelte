<!-- @license
  SPDX-FileCopyrightText: 2022 Severen Redwood <me@severen.dev>
  SPDX-FileCopyrightText: 2022 Keetley Rate <fridgeineyes@gmail.com>
  SPDX-License-Identifier: AGPL-3.0-or-later
-->

<p>Î»-calculus</p>
