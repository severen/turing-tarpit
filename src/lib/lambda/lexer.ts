/** @license
 * SPDX-FileCopyrightText: 2022 Severen Redwood <me@severen.dev>
 * SPDX-FileCopyrightText: 2022 Keetley Rate <fridgeineyes@gmail.com>
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import { SyntaxError } from "$lib/lambda/syntax";

/** A lexical token. */
export type Token = {
  kind: TokenKind;
  lexeme: string;
  position: number;
};

/** The kind of a Token. */
export enum TokenKind {
  /** A `位` character, or its ASCII stand-in, `\`. */
  Lambda,
  /** A right arrow symbol, `->`. */
  RArrow,
  /** A right bracket character, `(`. */
  LParen,
  /** A right bracket character, `)`. */
  RParen,
  /** An identifier. */
  Ident,
  /** A marker for the end of a file. */
  EOF,
}

/** Syntactically analyse, or lex, a potential 位-calculus term. */
export function lex(input: string): Token[] {
  return new Lexer(input).lex();
}

/** A lexer for 位-calculus programs. */
class Lexer {
  /** The input string to parse. */
  #input: string;
  /** The current character position of this lexer within the input string. */
  #position = 0;
  /** The character position of the start of the current token. */
  #start = 0;

  /** The lexically analysed tokens. */
  #tokens: Token[] = [];

  readonly #reIdent = /^\p{ID_Start}\p{ID_Continue}*'*/u;
  readonly #reWhitespace = /^\p{Pattern_White_Space}+/u;

  constructor(input: string) {
    this.#input = input;
  }

  /** Lex the input string to produce a stream of tokens. */
  lex(): Token[] {
    while (!this.#isAtEnd()) {
      this.#start = this.#position;
      this.#lexToken();
    }

    this.#tokens.push({ kind: TokenKind.EOF, lexeme: "", position: this.#position });
    return this.#tokens;
  }

  /** Lex a single token. */
  #lexToken(): void {
    const char = this.#advance();
    switch (char) {
      case "位":
      case "\\":
        this.#pushToken(TokenKind.Lambda);
        break;
      case "-": {
        if (this.#match(">")) {
          this.#pushToken(TokenKind.RArrow);
        } else {
          throw new SyntaxError(this.#start, "got invalid token -");
        }
        break;
      }
      case "(":
        this.#pushToken(TokenKind.LParen);
        break;
      case ")":
        this.#pushToken(TokenKind.RParen);
        break;
      default: {
        const tail = this.#input.slice(this.#start);
        if (this.#reIdent.test(char)) {
          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
          const ident = tail.match(this.#reIdent)![0];
          this.#position += ident.length - 1;
          this.#pushToken(TokenKind.Ident);
        } else if (this.#reWhitespace.test(char)) {
          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
          const whitespace = tail.match(this.#reWhitespace)![0];
          this.#position += whitespace.length - 1;
        } else {
          throw new SyntaxError(this.#start, `got invalid token ${char}`);
        }
        break;
      }
    }
  }

  #pushToken(kind: TokenKind): void {
    this.#tokens.push({
      kind,
      lexeme: this.#input.slice(this.#start, this.#position),
      position: this.#start,
    });
  }

  /** Get the current character and advance the lexer. */
  #advance(): string {
    return this.#input[this.#position++];
  }

  /** Get the current character without advancing the lexer. */
  #peek(): string {
    return this.#input[this.#position];
  }

  /** Advance the lexer if the current character equals the given character. */
  #match(char: string): boolean {
    if (this.#peek() === char) {
      this.#position += 1;
      return true;
    }

    return false;
  }

  /** Check if the lexer is at the end of the input. */
  #isAtEnd(): boolean {
    return this.#position === this.#input.length;
  }
}
