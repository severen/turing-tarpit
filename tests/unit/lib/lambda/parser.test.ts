/** @license
 * SPDX-FileCopyrightText: 2022 Severen Redwood <me@severen.dev>
 * SPDX-FileCopyrightText: 2022 Keetley Rate <fridgeineyes@gmail.com>
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import { describe, expect, it } from "vitest";

import { parse } from "$lib/lambda/parser";
import { mkAbs, mkApp, mkVar } from "$lib/lambda/syntax";

describe("Î»-calculus Parser", () => {
  it("parses variables", async () => {
    const vs = ["f", "g", "h", "x", "y", "z", "foo"];
    for (const v of vs) {
      expect(parse(v)).toEqual(mkVar(v));
    }
  });

  it("parses abstractions", async () => {
    expect(parse("\\x -> x")).toEqual(mkAbs("x", mkVar("x")));
  });

  it("parses nested abstractions", async () => {
    expect(parse("\\f -> \\x -> f")).toEqual(mkAbs("f", mkAbs("x", mkVar("f"))));
  });

  it("parses applications", async () => {
    expect(parse("f x")).toEqual(mkApp(mkVar("f"), mkVar("x")));
  });

  it("parses more applications", async () => {
    expect(parse("f x y")).toEqual(mkApp(mkApp(mkVar("f"), mkVar("x")), mkVar("y")));
  });

  it("parses even more applications", async () => {
    expect(parse("f (g x)")).toEqual(mkApp(mkVar("f"), mkApp(mkVar("g"), mkVar("x"))));
    expect(parse("f (\\x -> f x)")).toEqual(
      mkApp(mkVar("f"), mkAbs("x", mkApp(mkVar("f"), mkVar("x")))),
    );
  });

  it("does not parse the empty string", async () => {
    expect(() => {
      parse("");
    }).toThrowErrorMatchingSnapshot();
  });

  it("consumes the entire input", async () => {
    expect(() => {
      parse("this is a valid term but ; is not");
    }).toThrowErrorMatchingSnapshot();
  });
});
