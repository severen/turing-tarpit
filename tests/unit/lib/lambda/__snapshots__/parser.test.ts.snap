// Vitest Snapshot v1

exports[`λ-calculus Parser > consumes the entire input 1`] = `"expected EOF, got ;"`;

exports[`λ-calculus Parser > does not parse the empty string 1`] = `"unexpected EOF"`;
